components/layout/flex.rs

```diff
@@ -69,22 +73,195 @@ impl AxisSize {
+            // https://drafts.csswg.org/css-flexbox-1/#min-size-auto
+            Mode::Inline => {
+                let basis = from_flex_basis(self.style.get_position().flex_basis,
+                                            self.style.content_inline_size(),
+                                            Some(containing_length));
+
+                // These methods compute auto margins to zero length, which is exactly what we want.
+                block.fragment.compute_border_and_padding(containing_length,
+                                                          border_collapse::T::separate);
+                block.fragment.compute_inline_direction_margins(containing_length);
+                block.fragment.compute_block_direction_margins(containing_length);
+
+                let adjust_size = match self.style.get_position().box_sizing {
+                    box_sizing::T::border_box => {
+                        let margin = self.style.logical_margin();
+                        (MaybeAuto::from_style(margin.inline_start, Au(0)).specified_or_zero() +
+                         MaybeAuto::from_style(margin.inline_end, Au(0)).specified_or_zero())
+                    }
+                    box_sizing::T::content_box => block.fragment.surrounding_intrinsic_inline_size(),
+                };
+                let content_size = block.base.intrinsic_inline_sizes.preferred_inline_size - adjust_size;
+                self.base_size = basis.specified_or_default(content_size);
+                self.max_size = specified_or_none(self.style.max_inline_size(), containing_length)
+                    .unwrap_or(MAX_AU);
+                self.min_size = specified(self.style.min_inline_size(), containing_length);
+            },
+            Mode::Block => {
+                let basis = from_flex_basis(self.style.get_position().flex_basis,
+                                            self.style.content_block_size(),
+                                            Some(containing_length));
+                let content_size = match self.style.get_position().box_sizing {
+                    box_sizing::T::border_box => block.fragment.border_box.size.block,
+                    box_sizing::T::content_box => block.fragment.border_box.size.block
+                        - block.fragment.border_padding.block_start_end(),
+                };
+                self.base_size = basis.specified_or_default(content_size);
+                self.max_size = specified_or_none(self.style.max_block_size(), containing_length)
+                    .unwrap_or(MAX_AU);
+                self.min_size = specified(self.style.min_block_size(), containing_length);
+            }
+        }
+    }
+
+    /// Return the outer main size of the item, including paddings and margins,
+    /// clamped by max and min size.
+    pub fn outer_main_size(&self, mode: Mode) -> Au {
+        let ref fragment = self.flow.as_block().fragment;
+        let adjustment;
+        match mode {
```

Comment:
nit: you can shorten this to `let adjustment = match mode { ... }` and then remove the `adjustment =`s below and the semicolons